relayer/src/link.rs

```diff
@@ -376,188 +441,467 @@ impl RelayPath {
+            events.iter_mut().for_each(|ev| ev.set_height(new_height));
         }
 
         Ok(())
     }
 
-    fn reset_buffers(&mut self) {
-        self.dst_msgs_input_events = vec![];
-        self.src_msgs_input_events = vec![];
-        self.packet_msgs = vec![];
-        self.timeout_msgs = vec![];
-    }
-
-    fn relay_pending_packets(&mut self) -> Result<(), LinkError> {
-        println!("clearing old packets on {:#?}", self.channel);
+    fn relay_pending_packets(&mut self, height: Height) -> Result<(), LinkError> {
+        info!("[{}] Clearing old packets", self);
         for _i in 0..MAX_ITER {
-            if self.build_recv_packet_and_timeout_msgs().is_ok()
-                && self.send_update_client_and_msgs().is_ok()
-                && self.build_packet_ack_msgs().is_ok()
-                && self.send_update_client_and_msgs().is_ok()
+            if self
+                .build_recv_packet_and_timeout_msgs(Some(height))
+                .is_ok()
+                && self.build_packet_ack_msgs(Some(height)).is_ok()
             {
                 return Ok(());
             }
-            self.reset_buffers();
-            self.all_events = vec![];
         }
         Err(LinkError::OldPacketClearingFailed)
     }
 
-    pub fn clear_packets(&mut self, height: Height) -> Result<(), LinkError> {
+    /// Should not run more than once per execution.
+    pub fn clear_packets(&mut self, above_height: Height) -> Result<(), LinkError> {
         if self.clear_packets {
-            self.src_height = height
-                .decrement()
-                .map_err(|e| LinkError::Failed(e.to_string()))?;
-
-            self.relay_pending_packets()?;
+            let clear_height = above_height.decrement().map_err(|e| LinkError::Failed(
+                format!("Cannot clear packets @height {}, because this height cannot be decremented: {}", above_height, e.to_string())))?;
+            self.relay_pending_packets(clear_height)?;
+            info!("[{}] Finished clearing pending packets", self);
             self.clear_packets = false;
         }
 
         Ok(())
     }
 
-    /// Iterate through the IBC Events, build the message for each and collect all at same height.
-    /// Send a multi message transaction with these, prepending the client update
-    pub fn relay_from_events(&mut self, batch: EventBatch) -> Result<(), LinkError> {
+    /// Generate & schedule operational data from the input `batch` of IBC events.
+    pub fn process_event_batch(&mut self, batch: EventBatch) -> Result<(), LinkError> {
         self.clear_packets(batch.height)?;
 
-        // collect relevant events in self.all_events
-        self.collect_events(&batch.events);
-        self.adjust_events_height()?;
+        // Collect relevant events from the incoming batch & adjust their height.
+        let mut events = self.filter_events(&batch.events);
+        self.adjust_events_height(&mut events)?;
```

Comment:
I think we can get rid of this function. We can do on-chain updates at lower than latest client height now.

Reply:
Removing this uncovers an issue in `build_update_client()` where instead of picking a `trusting_height` as the highest consensus state height on chain we need to pick the the highest smaller than the `target_height`. So either we undo this change or wait for misbehaviour PR #691 that implemented an API to retrieve these heights.